### ISSUE TITLE
ESP responds with a 502 when verifying the signature of a JWT over a certain length

### ISSUE THREAD

**Issue by spopezen** [cloudendpoints/esp#538]
We've ran into an issue where requests with a Bearer token in the header, cause a 502 response from the ESP when the token is over a certain length (~3000 characters or so, could be less).

We've been able to track the bug introduction to the 1.21.0 release (likely this commit: https://github.com/cloudendpoints/esp/commit/d38f25b2e0a75598d505c9b2fca1e941ada3bde4)

I don't have much more information to provide at this time, and have not had the time to build the esp locally to test and debug further, but it should be fairly easy to reproduce.

**Comment by qiwzhang** [comment]
X-Endpoint-API-UserInfo is passed to the backend in the HTTP header.   With JWT claim this big,  either Nginx upstream rejects it with its limit,  or your backend app HTTP server rejects it.  

Could you verify if it is rejected by your app?

**Comment by spopezen** [comment]
We aren't having any problems on 1.20.0 with the same build of our app. 1.21.0 and 1.20.0 appear to be using the same nginx version, and I'm not seeing nginx config changes between the two releases, so it seems unlikely to be nginx. I do see that the endpoints-tools version changed, but have not yet looked at the diff between that version upgrade.

**Comment by qiwzhang** [comment]
Sorry,  I did not state it clearly.  

ESP 1.21 added JWT claim to the X-Endpoint-API-UserInfo which is added to the HTTP request send to the backend app.   The original JWT header is not removed in most cases, with the added header,  the header will be very big.  It may have reached Nginx buffer limit or your application buffer limit. 

You can help to debug it to see which limit is reached.  You can add --enable_debug flag in the ESP container flag, and check the ESP log inside the container in /var/log/nignx/error.log

**Comment by spopezen** [comment]
Ah, thanks for the clarification. Using --enable_debug flag, I was able to track the problem to my app.
Sorry for the noise, and thanks for the help! Closing this issue.

For people that may run into the same issue; node recently changed max request header size from 80k to 8k and ESP release 1.21 started adding the claims from an access token to the request header, which could push a request header size over that 8k limit (in certain cases) to an upstream node app.
